### PR TITLE
[common] Add ansi_filter

### PR DESCRIPTION
**Pull request by ljusten**
Adds a function to filter ANSI escape sequences from a string. Executing SSH commands on Windows yields output that is full of ANSI escape sequences if the "-tt" (forced TTY) argument is used. One particular escape sequence sets the window title to "c:\windows\system32\cmd.exe". This string is null terminated and messes with parsing the actual output later in that string. The filter function removes those escape sequences.

The outout is still a bit messed up, even after removing escape sequences. Some sequences delete rows and move the cursor. Without properly interpreting these sequences it doesn't seem possible to retrieve the proper output.

In a future CL the -tt argument is removed on Windows, which removes the necessity to filter ANSI codes. However, sometimes the target architecture is not known (yet), so that it is still useful to filter ANSI codes in that case to print useful debug output.

### CODE REVIEW

File: common/ansi_filter.h [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+#ifndef COMMON_ANSI_FILTER_H_
+#define COMMON_ANSI_FILTER_H_
+
+#include <string>
+
+namespace cdc_ft {
+namespace ansi_filter {
+
+// Removes ANSI escape sequences from |str|.
```
Comment (florian-kuebler) [inline common/ansi_filter.h:25]: Nit: The [style guide](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) suggests to not abbreviate by deleting letters within a word.

Comment (florian-kuebler) [inline common/ansi_filter.h:25]: In general, I would suggest to have the comment more descriptive:
* State the usage
* (pre/post conditions; invariants) -- not relevant here
* Give an example

At least, I would give an example.

See the [style guide](https://google.github.io/styleguide/cppguide.html#Function_Comments) for a detailed list.

File: common/ansi_filter.cc [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+constexpr uint8_t kOSCI = 0x9D;  // Operating System Command Introducer
+
+}  // namespace
+
+std::string RemoveEscapeSequences(std::string str) {
+  size_t write_idx = 0;
+  State state = State::kNotInSequence;
+
+  for (size_t read_idx = 0; read_idx < str.size(); ++read_idx) {
+    uint8_t ch = static_cast<uint8_t>(str[read_idx]);
```
Comment (florian-kuebler) [inline common/ansi_filter.cc:41]: In general, I would suggest to use fewer abbreviations (as suggested by the style guide)

File: common/ansi_filter.cc [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+        // Operating System Command.
+        if ((ch == kESC && next_ch == ']') || ch == kOSCI) {
+          read_idx += ch == kESC ? 1 : 0;
+          state = State::kOSC;
+          break;
+        }
+
+        // Char does not belong to control sequence.
+        str[write_idx++] = ch;
```
Comment (florian-kuebler) [inline common/ansi_filter.cc:69]: I think the code would be much more readable if you were not creating the string inplace, but rather iterating over the input string, appending characters while reading, and skipping the once inside a sequence.

File: common/ansi_filter.h (hunk 1 above)
Comment (ljusten) [reply common/ansi_filter.h:25]: Discussed offline, we're still using 80 chars, so it often hurts readability. We only use abbreviations for common stuff like str(ing), var(iable) etc.

File: common/ansi_filter.cc (hunk 3 above)
Comment (ljusten) [reply common/ansi_filter.cc:69]: Done.

File: common/ansi_filter.cc (hunk 2 above)
Comment (ljusten) [reply common/ansi_filter.cc:41]: Done.

File: common/ansi_filter.h (hunk 1 above)
Comment (ljusten) [reply common/ansi_filter.h:25]: Thanks for the suggestion, done!

File: common/ansi_filter.cc [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  for (size_t n = 0; n < input.size(); ++n) {
+    uint8_t ch = static_cast<uint8_t>(input[n]);
+    uint8_t next_ch =
+        static_cast<uint8_t>(n + 1 < input.size() ? input[n + 1] : 0);
+
+    switch (state) {
+      case State::kNotInSequence:
+        // Device Control String.
+        if ((ch == kESC && next_ch == 'P') || ch == kDCSI) {
+          n += ch == kESC ? 1 : 0;
```
Comment (florian-kuebler) [inline common/ansi_filter.cc:49]: As per [go/totw/3](https://goto.google.com/totw/3), consider using `StrAppend`.

Comment (florian-kuebler) [reply common/ansi_filter.cc:49]: Here and everywhere else.

File: common/ansi_filter.cc (hunk 2 above)
Comment (florian-kuebler) [reply common/ansi_filter.cc]: I think this is still open.

File: common/ansi_filter.cc (hunk 4 above)
Comment (florian-kuebler) [reply common/ansi_filter.cc:49]: Ups. Nevermind.

**Pull request merged**